src/annotation/annotation.test.ts

```diff
@@ -0,0 +1,435 @@
+  isClosed: false,
+};
+
+let blankAnnotation: Annotation;
+
+describe("Undo/Redo paintbrush tests", () => {
+  beforeEach(() => {
+    annotationsObject = new Annotations();
+    annotationsObject.addAnnotation("paintbrush");
+    [blankAnnotation] = annotationsObject.getAllAnnotations();
+  });
+
+  test("addBrushStroke / undo / redo", () => {
+    annotationsObject.addBrushStroke(brushStrokes[0]);
+    annotationsObject.addBrushStroke(brushStrokes[1]);
+    expect(annotationsObject.getAllAnnotations()[0].brushStrokes.length).toBe(
+      2
+    );
+
+    annotationsObject.undo();
+    expect(annotationsObject.getAllAnnotations()[0].brushStrokes.length).toBe(
+      1
+    );
+    const canUndoRedo = annotationsObject.undo();
+    expect(annotationsObject.getAllAnnotations()[0].brushStrokes.length).toBe(
+      0
+    );
+    expect(canUndoRedo).toStrictEqual({ undo: false, redo: true });
+    expect(annotationsObject.getAllAnnotations()).toStrictEqual([
+      blankAnnotation,
+    ]);
+
+    annotationsObject.redo();
+    expect(annotationsObject.getAllAnnotations()[0].brushStrokes.length).toBe(
+      1
+    );
+    annotationsObject.redo();
+    expect(annotationsObject.getAllAnnotations()[0].brushStrokes.length).toBe(
+      2
+    );
+  });
+
+  test("addBrushStrokeMulti / undo / redo", () => {
+    annotationsObject.addBrushStrokeMulti(brushStrokes);
+    expect(annotationsObject.getAllAnnotations()[0].brushStrokes).toStrictEqual(
+      brushStrokes
+    );
+
+    const canUndoRedo = annotationsObject.undo();
+    expect(canUndoRedo).toStrictEqual({ undo: false, redo: true });
+    expect(annotationsObject.getAllAnnotations()).toStrictEqual([
+      blankAnnotation,
+    ]);
+
+    annotationsObject.redo();
+    expect(annotationsObject.getBrushStrokeCoordinates(0)).toStrictEqual(
+      brushStrokes[0].coordinates
+    );
+    expect(annotationsObject.getBrushStrokeCoordinates(1)).toStrictEqual(
+      brushStrokes[1].coordinates
+    );
+  });
+
+  test("clearBrushStrokes / undo / redo", () => {
+    annotationsObject.addBrushStrokeMulti(brushStrokes);
+    expect(annotationsObject.getAllAnnotations().length).toBe(1);
+    expect(annotationsObject.getAllAnnotations()[0].brushStrokes.length).toBe(
+      2
+    );
+
+    annotationsObject.clearBrushStrokes();
+    expect(annotationsObject.getAllAnnotations()[0].brushStrokes).toStrictEqual(
+      []
+    );
+
+    // test undo:
+    let canUndoRedo = annotationsObject.undo();
+    expect(canUndoRedo).toStrictEqual({ undo: true, redo: true });
+    const annotations = annotationsObject.getAllAnnotations();
+    expect(annotations.length).toBe(1);
+    expect(annotations[0].brushStrokes.length).toBe(2);
+
+    // test redo:
+    canUndoRedo = annotationsObject.redo();
+    expect(canUndoRedo).toStrictEqual({ undo: true, redo: false });
+    expect(annotationsObject.getAllAnnotations()[0].brushStrokes).toStrictEqual(
+      []
+    );
+  });
+
+  test("clickSelect / undo / redo", () => {
```

Comment:
Does the undo/redo functionality depend on `clickSelect`? If it doesn't, I wouldn't bother testing the two together.

Reply:
I'm testing that clickSelect itself can be undone / redone. I'm doing so with all methods where undo/redo is applicable - test both the method itself and that it can be undone / redone. Since they all have separate code for computing the inverse action and adding it to the undo stack, undo needs to be tested separately for each.

Reply:
I didn't know click-to-select could be undone and redone.